Install newest Chrome browser before running tests

clarmso opened the pull request:
<!-- 
Thanks for contributing!
Read our contribution guidelines here: 
https://github.com/cypress-io/cypress/blob/develop/.github/CONTRIBUTING.md 
-->

I'm trying to investigate some Chrome 79+ related issues from #6023 by running the existing tests using the latest Chrome browser.  We've been lagging in using the latest stable version of Chrome in testing Cypress itself.

I've some thoughts on how to use the latest stable version of Chrome in the testing.  This work would enable the Cypress developers to address the compatibilities between Cypress and Chrome in a timely manner.

If it all goes well, we may close #5236. 👍 

- Closes #5236

### User facing changelog

Add the ability to test the changes against the latest stable version of the Google Chrome browser.

### Additional details

The latest stable version of the Chrome is installed prior to running the test runs that uses Chrome.  This way, we don't need to update the Docker image indicated in this line right away when a new version of Chrome becomes available.

https://github.com/cypress-io/cypress/blob/639ba3206140756285d4b9043eda799eec04da61/circle.yml#L36

Since this Docker image has the repository of Chrome configured, we can just run apt to retrieve the latest version of Chrome.

### How has the user experience changed?

There should be no change in the user's experience.  The developers can automatically run the tests against the latest stable version of Google Chrome instead of an older one.

### PR Tasks

N/A

clarmso commented:
I am not authorized to run the server-performance-tests and all server-e2e-tests-* on CircleCI.  I cannot test these items on my end.  Could someone from Cypress help out?

https://app.circleci.com/github/cypress-io/cypress/pipelines/fa6c0831-9abb-43fb-b6b6-ef56a8322ee5/workflows/b2bed0f6-5601-4953-95b0-ec03ae4dcc19

driver-integration-tests-chrome installs and uses Chrome 79 successfully.
https://app.circleci.com/jobs/github/cypress-io/cypress/227689

clarmso commented:
Hm...Let me install the latest Chrome browser when it is needed.  The installation should occur in the server-e2e-tests-chrome-* and the driver-integration-tests-chrome.

I'm not 100% with the solution because Chrome is (re)installed in every server-e2e-tests-chrome-*.  I cannot just cache the .deb file for the Chrome installer because the browser may have other dependencies in the future. "apt-install" is the sure way to install Chrome.  In addition, the Chrome installer is fetched from an external server. We really need the connection between CircleCI and the location of the Chrome debian package and its dependencies' to work consistently.

If there's a more efficient solution, please let me or the maintainers know!

clarmso commented on circle.yml:
```diff
@@ -66,6 +74,10 @@ commands:
     steps:
       - attach_workspace:
           at: ~/
+      - when:
```
I intended to have the newest Chrome not to be installed for the server-e2e-tests-electron-*.  install-latest-chrome is still installed for the electron tests. :(  I could add a boolean flag to dictate whether the browser is installed or not.  Is there a better way?

flotwig replied:
I wonder if you could update the `install-latest-chrome` command to exit if the supplied `browser` is not `chrome`, maybe something like:

```sh
if [[ << parameters.browser >> = 'chrome' ]]
then
	echo 'Running Chrome tests, installing latest version...'
else
	echo 'Running in << parameters.browser >>, not updating Chrome...'
	exit 0
fi
```

flotwig commented on circle.yml:
```diff
@@ -55,6 +55,14 @@ executors:
+    steps:
+      - run:
+          command: |
+            apt update ;
```
`apt-get` is usually recommended for scripts, but this is probably fine too since we aren't depending on the output

flotwig commented on circle.yml:
```diff
@@ -55,6 +55,14 @@ executors:
+            apt update ;
+            apt install google-chrome-stable -y ;
+            which google-chrome ;
+            google-chrome --version        
```
maybe add an `echo 'Google Chrome version:'` above this to help break it out

clarmso commented:
I've addressed the issues and the suggestions above.  The tests driver-integration-tests-chrome installs the newest stable Chrome successfully, but some tests failed (https://app.circleci.com/jobs/github/cypress-io/cypress/239271).  Let me investigate if they're Chrome 79 specific.  Those failures should not be related to this change in circleci.yml.

The electron case as seen in server-e2e-tests-electron-* is not tested yet.

clarmso commented:
The test failure is from `packages/driver/integration/commands/actions/type_spec.js` with the following error message:
```
  1) src/cy/commands/actions/type user experience can print table of keys on click:
     TypeError: Cannot convert undefined or null to object
      at Function.keys (<anonymous>)
      at findReactInstance (cypress:///./cypress/support/utils.js:18:20)
      at withMutableReporterState (cypress:///./cypress/support/utils.js:46:24)
      at Context.eval (cypress:///./cypress/integration/commands/actions/type_spec.js:4381:16)
```
I haven't been able to reproduce this error locally either by running the single test in isolation or all tests  from the file.  I've Chrome 79.0.3945.130 installed on my computer as well. I don't like intermittent test failures, but let me kick off the tests again through rebasing.

clarmso commented on circle.yml:
```diff
@@ -61,10 +61,10 @@ commands:
         description: browser shortname to target
-        type: enum
-        enum: [ "chrome", "electron" ]
+        type: string
```
I changed the `browser` to enum because of more browsers are coming. :smile: Either way is fine.

Please support more browsers in the near future! :bowing_woman:

flotwig replied:
#1359 is super close :fox_face:

jennifer-shehane commented:
Opened a new issue to revert / change implementation of this due to newly exhibited problems. https://github.com/cypress-io/cypress/issues/6328
